index.js

```diff
@@ -1,41 +1,113 @@
 
-module.exports = function(opt) {
-    opt = xtend({ 
-        width: 1, 
+
+module.exports = function (shader, opt) {
+    if (!opt) {
+        //just options
+        if (typeof shader === 'object' && !shader.fragShader) {
+            opt = shader
+        }
+        //just a shader
+        else {
+            opt = {
+                shader: shader
+            }
+        }
+    }
+    else {
+        opt.shader = shader
+    }
+
+    opt = xtend({
+        width: 1,
         height: 1,
-        preserveDrawingBuffer: true 
+        preserveDrawingBuffer: true,
+        float: true
     }, opt)
-    
-    var gl = opt.gl || create(opt)
+
+    var gl = opt.shader && opt.shader.gl || opt.gl || create(opt)
     if (!opt.shader)
         throw new Error('no shader supplied to gl-shader-output')
-    
+
+    //set gl context dims
+    gl.canvas.width = opt.width
+    gl.canvas.height = opt.height
+
     var shader = typeof opt.shader === 'function'
             ? opt.shader(gl)
             : opt.shader
 
+    //create gl-shader, if only fragment shader source
+    if (typeof shader === 'string') {
+        shader = Shader(gl, '\
+            attribute vec2 position;\
+            void main() {\
+              gl_Position = vec4(position, 1.0, 1.0);\
+            }\
+        ' , shader)
+    }
+
+    //micro optimizations
+    gl.disable(gl.DEPTH_TEST)
+    gl.disable(gl.BLEND)
+    gl.disable(gl.CULL_FACE)
+    gl.disable(gl.DITHER)
+    gl.disable(gl.POLYGON_OFFSET_FILL)
+    gl.disable(gl.SAMPLE_COVERAGE)
+    gl.disable(gl.SCISSOR_TEST)
+    gl.disable(gl.STENCIL_TEST)
+
+    var buffer = gl.createBuffer()
+    gl.bindBuffer(gl.ARRAY_BUFFER, buffer)
+    gl.bufferData(gl.ARRAY_BUFFER, new Float32Array([-1, -1, -1, 3, 3, -1]), gl.STATIC_DRAW)
+    shader.attributes.position.pointer()
+
+    //try to use floats
+    var float = glExt(gl).OES_texture_float && opt.float
+
+    //set framebuffer as a main target
+    var framebuffer = new Framebuffer(gl, [opt.width, opt.height], {
+        float: float,
+        depth: false,
+        color: 1
+    })
+    framebuffer.bind()
+    shader.bind()
+
+
     function process(uniforms) {
+        var w = gl.drawingBufferWidth, h = gl.drawingBufferHeight
+
         gl.clearColor(0, 0, 0, 0)
         gl.clear(gl.COLOR_BUFFER_BIT)
 
-        shader.bind()
-
         //if user specifies some uniforms
-        if (uniforms)
+        if (uniforms) {
+            shader.bind()
             assign(shader.uniforms, uniforms)
+        }
 
         //full-screen quad
-        triangle(gl)
+        gl.drawArrays(gl.TRIANGLES, 0, 3);
+
+        if (float) {
+            var pixels = new Float32Array(w * h * 4)
+            gl.readPixels(0, 0, w, h, gl.RGBA, gl.FLOAT, pixels)
+        }
+        else {
+            var pixels = new Uint8Array(w * h * 4)
+            gl.readPixels(0, 0, w, h, gl.RGBA, gl.UNSIGNED_BYTE, pixels)
+            pixels = new Float32Array(pixels).map(function(p) {
```

Comment:
`Float32Array#map` doesn't work in Safari and probably some other older browsers. :smile: